### PR TITLE
Update MSFT_AADGroup.psm1

### PR DESCRIPTION
**Pull request by ifinch**
#### Pull Request (PR) description
line 1050
changed from if ($Filter -like "*endsWith*") to if ($Filter -like "*endsWith*" -or $Filter -like "*onPremisesSyncEnabled eq null*" -or $Filter -like "*onPremisesSyncEnabled ne true*")

#### This Pull Request (PR) fixes the following issues
it resolves the issue in #3935

### CODE REVIEW

**Comment by ricmestre** [comment]
I added that condition to the $Filter to itch a scratch at the time but in order to actually solve this (also for other resources that support filtering) someone needs to go through the list in https://learn.microsoft.com/en-us/graph/aad-advanced-queries?tabs=http#device-properties and implement the condition correctly.

Looking at that table it shows that for your use case you only need to define *ConsistencyLevel* and *$count* if comparing *onPremisesSyncEnabled* to null, it's not actually required if comparing to true or false, I've just tested it in my tenant and can verify this so your patch is half incorrect.

The best way would be to have the MSGraph SDK decide which properties need *ConsistencyLevel* and *$count* and add them automatically, this has been requested several years ago but it's still not in place, see https://github.com/microsoftgraph/msgraph-sdk-powershell/issues/513.

**Comment by ifinch** [comment]
changed if ($Filter -like "endsWith" -or $Filter -like "onPremisesSyncEnabled eq null" -or $Filter -like "onPremisesSyncEnabled ne 
true") to be if ($Filter -like "endsWith" -or $Filter -like "onPremisesSyncEnabled eq null")  removing "onPremisesSyncEnabled ne true"

**Comment by ifinch** [comment]
added the block to check allowed advanced Group properties from the table https://learn.microsoft.com/en-us/graph/aad-advanced-queries?tabs=http#group-properties that support eq null.

start line 1050, ends 1078

# Define the list of attributes
        $attributesToCheck = @(
            "description",
            "displayName",
            "hasMembersWithLicenseErrors",
            "mail",
            "mailNickname",
            "onPremisesSecurityIdentifier",
            "onPremisesSyncEnabled",
            "preferredLanguage"
        )

        # Initialize a flag to indicate whether any attribute matches the condition
        $matchConditionFound = $false

        # Check each attribute in the list
        foreach ($attribute in $attributesToCheck) {
            if ($Filter -like "*$attribute eq null*") {
                $matchConditionFound = $true
                break
            }
        }

        # If any attribute matches, add parameters to $ExportParameters
        if ($matchConditionFound) {
            $ExportParameters.Add('CountVariable', 'count')
            $ExportParameters.Add('ConsistencyLevel', 'eventual')
        }

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADGroup/MSFT_AADGroup.psm1 [hunk 1]
```diff
@@ -1047,10 +1047,36 @@ function Export-TargetResource
...
+        # Check each attribute in the list
+        foreach ($attribute in $attributesToCheck) {
+            if ($Filter -like "*$attribute eq null*") {
+                $matchConditionFound = $true
+                break
+            }
+        }
+
+        # If any attribute matches, add parameters to $ExportParameters
+        if ($matchConditionFound) {
```
Comment (ricmestre) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADGroup/MSFT_AADGroup.psm1:1075]: Now you're missing *endsWith* here, this is not an attribute but a function that you can call on the query which also needs to be checked and if you remove it then it stops working.

Comment (ifinch) [reply Modules/Microsoft365DSC/DSCResources/MSFT_AADGroup/MSFT_AADGroup.psm1]: changed the condition to look as below to take into account endsWith
so it is now

if ($matchConditionFound -or $Filter -like "*endsWith*"))

**Comment by ricmestre** [comment]
@NikCharlebois I'm not 100% sure but from the tests I've seen always including *ConsistencyLevel* and *CountVariable* in the cmdlets while filtering doesn't cause any issues, I might be wrong here though but if that's the case then adding them always if *$Filter* is not null would be better then have to chase whatever is mentioned in [0]. As an example they even omit the function *endsWith* that I added so I don't know exactly to which properties it applies so I had to apply the big hammer to the condition which now I asked @ifinch to change.

Any insights?

[0] https://learn.microsoft.com/en-us/graph/aad-advanced-queries?tabs=http#device-properties

**Comment by NikCharlebois** [comment]
Can we please add an entry to the changelog? Thanks

**Comment by NikCharlebois** [comment]
> _ConsistencyLevel_

Not sure this is really something we want to push at scale since it can change the current behavior in certain scenarios:
https://devblogs.microsoft.com/microsoft365dev/build-advanced-queries-with-count-filter-search-and-orderby/#what-is-eventual-consistency

**Comment by NikCharlebois** [comment]
Can you update your branch? Not sure why unit tests are failing in your PR. Thanks

**Comment by ifinch** [comment]
Done

**Comment by ifinch** [comment]
It is pending anything on my side?

**Pull request merged**